Add apply_filter attribute to recorder.purge service

cdce8p opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds the `apply_filter` attribute to the `recorder.purge` service. The attribute indicates that the purge service should remove all filtered states and events even if they occurred between `now` and `now - purge_days`. This feature is intended to be used by a user only, so there is no change to the daily purge.

## Rational / Use case
I recently noticed that my production database had quadrupled in size after I missed excluding a few entity_ids. Since I use a Raspberry PI with an SD card, this caused my whole system to crash. After updating my config, I ran a manual purge, to try to reduce the db size. I noticed then, that the filter is only applied when a state or event is being added. I could have used `keep_days: 0` or removed the database completely but I just wanted to remove the falsely added entries. I had to remove them manually.

With this change, a user can just run `recorder.purge` with `repack: true` and `apply_filter: true` to remove the entries without loosing their entire history or knowing SQL.

## Implementation details
- `apply_filter: true` has to be set in a manual service call to `recorder.purge`. As described earlier, the daily purge doesn't change.
- To check which `states` should be removed, I used `distinct(States.entity_id).all()` and check for each entity if it should be filtered. For events I used `distinct(Events.event_type).all()`.
- Only states and events between `now` and `now - purge_days` are considered. Everything before then will be removed by the existing logic.
- If a state is to be removed, I also remove the event with the matching `event_id`. (`state_changed` events)
- If an event is to be removed, I also remove the matching state. (for `state_changed` events)
- Similar to the existing purge, the deletion happens in small increments, currently set to one hour. If the purge isn't finished, it will return False, which calls the purge method again.
- I have added extensive integration tests to validate the change.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
recorder:
  exclude:
    entities:
      - sensor.purge
    event_types:
      - timer_out_of_sync


# Service call
recorder.purge:
  apply_filter: true
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16377

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cdce8p committed:
Add apply_filter option to recorder.purge service

* Always delete states, even if event_id is None
* Remove subsets
* Set old_state_ids to Null before removing states

cdce8p committed:
Added tests

bdraco commented on homeassistant/components/recorder/purge.py:
```diff
@@ -101,3 +125,174 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+            event_ids_subset = event_ids[i : i + subset_size]
+            deleted_rows_events += (
+                session.query(Events)
+                .filter(Events.event_type == EVENT_STATE_CHANGED)
+                .filter(Events.time_fired < batch_purge_before)
```
```suggestion
                .filter(Events.event_type == EVENT_STATE_CHANGED)
                .filter(Events.time_fired < batch_purge_before)
```

If we have the specific event ids, do we need this?

cdce8p replied:
I included those filters for query optimization. AFAIK the `in_` check is really inefficient. It compares each `event_id` with each one in the list. The two checks above are executed beforehand and thus limit which `event_ids` are even checked.

bdraco replied:
Makes sense 👍

bdraco replied:
Maybe a subquery to get the event ids ?

cdce8p replied:
I want to limit the number of max comparisons. See https://github.com/home-assistant/core/pull/46678#discussion_r577346531 for the detailed explanation

cdce8p replied:
> Maybe a subquery to get the event ids ?

I think a subquery wouldn't work here, since the states need to be deleted first.

bdraco commented:
Out of time to fully run though this. It going to make it more difficult for someone to fix https://github.com/home-assistant/core/issues/44958 

I might need to bite the bullet on this and get a MSSQL server setup so I can make the change to fix that before we merge this since it doesn't look like we have a MSSQL dev stepping up to do the work in that issue.

bdraco commented:
I've made the changes to fix MSSQL.  Hopefully it will be a small change to make the change sets compatible

https://github.com/home-assistant/core/pull/46678/files

cdce8p commented:
> I've made the changes to fix MSSQL. Hopefully it will be a small change to make the change sets compatible
> 
> https://github.com/home-assistant/core/pull/46678/files

I'll check later how to best update my PR. Not sure if that will be a small change just yet.

cdce8p commented:
@bdraco I've updated the PR to comply with the changes from #44958 and extended the test cases. It's ready for review.

bdraco commented:
Did you get a chance to test this with MySQL?

cdce8p commented:
> Did you get a chance to test this with MySQL?

It took me a  moment to set it up, but it works as intended. I also tested it with `ON DELETE NO ACTION` set for `old_state_id`.

bdraco commented on homeassistant/components/recorder/purge.py:
```diff
@@ -101,3 +125,193 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+        event_id
+        for (event_id,) in session.query(States.event_id)
+        .filter(States.last_updated < batch_purge_before)
+        .filter(States.event_id is not None)
```
This shouldn't be possible.  Did you find a row like this?

```suggestion
        .filter(States.event_id is not None)
```

cdce8p replied:
No I haven't. That was one case where I wasn't sure if it is implied. Technically it would be possible since the schema does allow it (`event_id` doesn't have `nullable=false`).
Personally I would like to leave it in if that's alright with you. Performance wise this shouldn't matter.

bdraco commented:
Looks good 👍 

I want to get another set of 👀 on https://github.com/home-assistant/core/pull/46678 and then they should both be good to merge

cdce8p commented:
> Looks good 👍
> 
> I want to get another set of 👀 on #46678 and then they should both be good to merge

This PR is compatible with with both, the current implementation and #46678. So it should be possible to merge it earlier. I would even suggest to do it that way. The other one is a bit easier to rebase. Should be just adding one indent for the change in `purge.py`.

bdraco commented:
Merging this on first makes sense as the other one will be easier to fix conflicts.

I've asked for additional reviews.

cdce8p committed:
Improve logic

cdce8p commented:
@bdraco I've rebased and updated this PR. The first two commits include everything from before, including the tests. In the last commit (https://github.com/home-assistant/core/pull/45826/commits/3961b07f0901402fb4019d38bfd6d7b398ea8640), I've updated the logic which should now be much cleaner and easier to understand, in part thanks to your refactoring in #46678. The PR is ready for review now.

bdraco commented:
Looks good. I'll do some testing with it as soon as I have some free cycles

PeteBa commented:
Sorry to come late - feel free to ignore - but would a `purge_entities` service call taking an array of entities be a viable alternative here ? Appreciate they would both get to the same outcome but `purge_entities` might be more discoverable and cover a broader set of use cases.

I had a similar experience to OP when I tried out an integration that I decided not to go with and then needed to get rid of the unwanted entities. In this use-case, the original proposal would mean editing the recorder yaml, restarting and running `purge(apply_filters: true)` service vs just running `purge_entities(entities)`.

cdce8p commented:
@PeteBa That might be an idea, although I would imagine that it might get abused in the end. Just as an example, say a user as a default retention period of 7 days but decides to use service calls to remove just specific entities earlier, maybe even with repacking each time. That's additional wear on their storage medium, an SD card in the worst case, and system slow downs.

Using the `recorder.exclude` is much more deliberate, you need to modify your config and restart after all. You also don't have the issue of forgetting to add the entity_id to the exclude list if you continue to use the integration.

IMO this feature isn't intended for regular use. It works as you would expect but I would only recommend it if you really need it.

@bdraco What do you think?

PeteBa commented:
@cdce8p , thanks for the response ! and I agree with the "only in emergency" usage. Also, with the comment that the `apply_filter` approach provides that additional opportunity for consideration. I'm not sure that one is more open to abuse as the functionality is near identical and accessible once awareness gets around. Stripping it down, I would suggest `purge_entities` is a more atomic building block while `apply_filters` addresses a more specific use-case. Both reasonable.

Either way - and more importantly ! - the functionality in the PR is great and would have saved me a ton of grief a few weeks ago as I was trying to figure out SQL 8)

bdraco commented:
> @bdraco What do you think?

It sounds like a great followup PR that @PeteBa could take on once this is merged since it will provide some great building blocks to enable that.

cdce8p commented:
> > @bdraco What do you think?
> 
> It sounds like a great followup PR that @PeteBa could take on once this is merged since it will provide some great building blocks to enable that.

I agree, it should almost be trivial with https://github.com/home-assistant/core/blob/3961b07f0901402fb4019d38bfd6d7b398ea8640/homeassistant/components/recorder/purge.py#L178-L179
The only thing left would be to decide how to best wire it up. It might make sense to create a new service for it. Maybe it could take a list of entities or an entity filter as argument? That way a user could still use the glob patterns if they want.

bdraco commented:
Tested and working as expected

bdraco reviewed (approved):
LGTM, test pass 👍 

Thanks!
